hpy/devel/src/runtime/hlist.c

```diff
@@ -0,0 +1,135 @@
+/**
+ * A manager for a list of HPy handles, allowing handles to be tracked
+ * and closed as a group.
+ *
+ * Note::
+ *    HList always keeps space for one extra handle free so that
+ *    HList_Track can always store the handle being passed to it,
+ *    even if it fails to automatically create space for future
+ *    handles. This allows HList_Free to close all handles passed to
+ *    HList_Track.
+ *
+ *    As a consequence, the minimum size for an HList is one handle.
+ *
+ * Example usage (inside an HPyDef_METH function)::
+ *
+ * long i;
+ * HPy key, value;
+ * HList hl;
+ *
+ * hl = HList_New(ctx);  // track the key-value pairs
+ * if (hl == NULL)
+ *     return HPy_NULL;
+ *
+ * HPy dict = HPyDict_New(ctx);
+ * if (HPy_IsNull(dict))
+ *     goto error;
+ *
+ * for (i=0; i<5; i++) {
+ *     key = HPyLong_FromLong(ctx, i);
+ *     if (HPy_IsNull(key))
+ *         goto error;
+ *     if (HList_Track(ctx, hl, key) < 0)
+ *         goto error;
+ *     value = HPyLong_FromLong(ctx, i * i);
+ *     if (HPy_IsNull(value)) {
+ *         goto error;
+ *     }
+ *     if (HList_Track(ctx, hl, value) < 0)
+ *         goto error;
+ *     result = HPy_SetItem(ctx, dict, key, value);
+ *     if (result < 0)
+ *         goto error;
+ * }
+ *
+ * success:
+ *    HList_Free(ctx, hl);
+ *    return dict;
+ *
+ * error:
+ *    HList_Free(ctx, hl);
+ *    HPy_Close(ctx, dict);
+ *    HPyErr_SetString(ctx, ctx->h_ValueError, "Failed!");
+ *    return HPy_NULL;
+ */
+
+#include <Python.h>
+#include "hpy.h"
+
+#define _HLIST_INITIAL_SIZE (5 + 1)
+
+struct _HList_s {
+    HPy_ssize_t size;
+    HPy_ssize_t next;
+    HPy *handles;
+};
+
+
+HPyAPI_RUNTIME_FUNC(HList)
+HList_New(HPyContext ctx)
```

Comment:
maybe we could add an optional param which specifies the initial size of the pool?

Reply:
I was thinking about this but was trying to avoid adding more va_list things. Perhaps I can just add a second function for creating the tracker.

Reply:
Second function added.